### PR TITLE
Minimap: Add 'map' mod with mapping kit item to enable minimap 

### PR DESCRIPTION
**Pull request by paramat**
Minimap: Add 'map' mod with mapping kit item to enable minimap

In survival mode, use of the minimap requires the mapping kit item.
Minimap radar mode is always disabled in survival mode.

Minimap and minimap radar mode will be automatically enabled in creative mode
and for any player with the 'creative' privilege.

The 'map.enable()' function is global so can be redefined by a mod for
alternative behaviour.
////////////////

screenshot_20171001_171753

mapping_kit2_ind128

Once crafted, to instantly enable, 'use' (left click) the item, otherwise the
minimap will be automatically enabled a few seconds later.

In the recipe:
Glass over steel ingot over wood is the compass.
Stick over steel ingot is a wooden quill. Black dye represents ink.

Compass needle tips are painted red for N and white for S as is standard.

### CODE REVIEW

**Comment by TeTpaAka** [comment]
Sure, take it with CC-BY-SA 3.0.

**Comment by Fixer-007** [comment]
>If a player loses the map item then the minimap stays usable for the rest of the game session but of course is disabled on next game session.

No go

**Comment by paramat** [comment]
Yeah, i'll add a checking function.

**Comment by paramat** [comment]
Added regular check, license and README.

**Comment by paramat** [comment]
There was a request for an option to enable minimap for survival players.
In survival using minimap only makes sense if you have the item, so you can use the 'give initial stuff' mod API to give the map to new players.
For existing players, i have made the checking function global and overridable, make a mod that depends on 'map' that contains:
```
function map.check_for_map(player)
		player:hud_set_flags({minimap = true})
end
```
This way you can alter behaviour in any way you want.

**Comment by paramat** [comment]
I might improve the map texture, it's too much like a treasure map currently.

**Comment by paramat** [comment]
I'll make the map stack max = 1 soon.

File: mods/map/init.lua [hunk 1]
```diff
@@ -0,0 +1,66 @@
+map = {}
+
+
+-- Global function to allow overriding
+
+function map.check_for_map(player)
+	if creative.is_enabled_for(player:get_player_name()) or
```
Comment (rubenwardy) [inline mods/map/init.lua]: I suggest making this an optional dependency

File: mods/map/init.lua [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+local timer = 0
+
+minetest.register_globalstep(function(dtime)
+	timer = timer + dtime
+	if timer < 7 then
+		return
+	end
+
+	timer = 0
+	for _, player in ipairs(minetest.get_connected_players()) do
```
Comment (rubenwardy) [inline mods/map/init.lua]: It's a common idiom to use minetest.after to run something at an interval

**Comment by TeTpaAka** [comment]
How about this one?
map_map

**Comment by GreenXenith** [comment]
> How about this one?

looks more like a picture than a map...

**Comment by paramat** [comment]
> The fact that dropping the map won't remove the ability to use the minimap is an issue.

Dropping the map does remove the ability.

File: mods/map/init.lua (hunk 1 above)
Comment (paramat) [inline mods/map/init.lua]: Ok. I'll add in an alternative direct getting of creative setting.

File: mods/map/init.lua (hunk 2 above)
Comment (paramat) [inline mods/map/init.lua]: Ok that could be better than always running a function every globalstep.

**Comment by DS-Minetest** [comment]
>Dropping the map does remove the ability after a few seconds, which is fine, the check searches inventory of all players so shouldn't be too frequent, i can reduce it to 5s.

I suggest also adding an `on_drop` function.

**Comment by GreenXenith** [comment]
I used paramat's original texture and modified it a bit.
map_map

**Comment by paramat** [comment]
A treasure map is what i'm trying to avoid, TeTpaAka's new one is close to what i intend.

This might as well wait for the egine radar HUD flag PR so i can also add the special remote-viewing object.

**Comment by GreenXenith** [comment]
> A treasure map is what i'm trying to avoid, TeTpaAka's new one is close to what i intend.

Fine with me, but because I still think his looks too much like a photo on some copy paper, I made TePpaAka's into a scroll version (unless it's the scroll part of the treasure map you are trying to avoid :P ).

map

**Comment by paramat** [comment]
https://github.com/minetest/minetest/pull/6273 merged.

**Comment by paramat** [comment]
Updated to use radar HUD flag, but i still have to test. New screenshots coming.

**Comment by Ekdohibs** [comment]
I'd suggest that the subterranean eye would activate radar for some limited amount of time, taking some wear, given that it is quite powerful. The amount of time and of wear need however to be chosen.

**Comment by Fixer-007** [comment]
map:subterranean_eye recipe is too easy?

**Comment by paramat** [comment]
Indeed i'm open to suggestions for how to enable radar mode.
I'm quite happy with the recipe for the surface map, just need a better texture.

**Comment by C1ffisme** [comment]
Just as a thought, I really don't like it when a Minecraft or Minetest mod has an expensive item that just requires a whole bunch of expensive raw materials. It feels less like I'm crafting things with my inventive skills and handiwork and more like I'm using an oddly designed in-game item shop.

Instead of using a bunch of expensive items, why not try either:
* Placing the expensive item or it's crafting components in a rare dungeon or other spawned structure OR
* Making a new raw material that can be used mainly for this, as well as a few other uses.

Speaking of new raw materials, it is worth noting that Tin is pretty unused at the moment. (Hint hint...)

**Comment by paramat** [comment]
I'm thinking perhaps we should have just one map item, which is then 'powered-up' somehow to also do radar. Question is, how do we turn this map into a magickal map? It needs to be earned somehow but in an interesting way.

**Comment by C1ffisme** [comment]
@paramat If we could attach items in the player's inventory to a clock, we could have a system where the player has to use a certain item to charge the map for a certain amount of time. (I feel like we need a magic item in the game that's relevant to vision. Minecraft has Spider Eyes and Ender Eyes, but we don't have a lot of mobs in the game. None, in fact.)

**Comment by GreenXenith** [comment]
> It needs to be earned somehow but in an interesting way.

There are so many things that need to happen before stuff like this is achieved easily. We currently have no experience system, no mobs, no enchanting, no realms, not even dungeon loot. Just using craft items isn't very interesting... perhaps dungeon loot would be a good idea (though a separate PR should be made for loot in general and then add the map to it with this PR). It would finally give players reason to explore dungeons (because right now they are completely and utterly useless - unless you are weird and like to make bases in them).

**Comment by paramat** [comment]
> There are so many things that need to happen before stuff like this is achieved easily.

No.
Dungeon loot is a good idea and is likely to happen.

**Comment by Fixer-007** [comment]
subterranean_eye should probably be even more expensive and with limited amount of activations (like 10)?

**Comment by GreenXenith** [comment]
> 
>>There are so many things that need to happen before stuff like this is achieved easily.
>
>No.

What I meant by "easliy" is rather that things can be added to the game without having to figure out a unique way for every single things (dungeon loot can produce many different things, enchanting could obtain different things, etc..). Not that it can't be done without adding stuff :P

**Comment by TumeniNodes** [comment]
I worked on a map texture for the heck of it, it's here if you like it. (might look better rotated onto it's side, idk, it's a toss-up)
map 01

**Comment by paramat** [comment]
Will try the texture.
That map mod is unsuitable because it doesn't limit the map to what is already explored, something that the engine could probably do and faster, by using the pixels created by the minimap. So i think best done as an extension of the minimap.

**Comment by TumeniNodes** [comment]
From memory (which is risky for me) it used to only map out areas you have explored. The map would change if you logged out, then back in again. There is the "remap" button but it could take forever to update or crash the game.
Either way, adding some means of being able to click into a full map page is really nice to have, especially for planning out areas for building. It is also very useful for marking out areas.
There was also a program you could use to create an out of game map.

Adding something like this as an extension to minimap would be desirable.

**Comment by paramat** [comment]
Ok sorry, i only skimmed the thread. Still, needs to be created by the minimap code that already creates local map, not using the python minetestmapper.

**Comment by jastevenson303** [comment]
I like this one too!  (https://github.com/jastevenson303/dcbl/commit/1a482f7c7c21a526f1406f812f640940b63bea25)

That seven second lag though!
: )

**Comment by paramat** [comment]
'Using' the item will immediately enable the feature, so the 7s delay is only an issue when you no longer have the item.

**Comment by paramat** [comment]
I'm still enthusiastic abou tthis PR, just considering the items and textures and what is necessary to get radar.

**Comment by jastevenson303** [comment]
It's a good thought about the strong recipes.  I initially thought "holy shit!" but I have to try it from fresh world/player to see.  Gives the player something to look forward to. ; )

**Comment by TumeniNodes** [comment]
> just considering the items and textures and what is necessary to get radar.

cough... and a clickable, pop-up window map... cough cough

**Comment by paramat** [comment]
Yes we want that, maybe later, and that woul dbe added to this mod.

**Comment by paramat** [comment]
TumeniNodes i'll probably use your map texture, maybe with a minor change that i'll ask your opinion on.

I'm so unsure about how to enable radar mode i think that can be done in a later PR, for now we can just have the map item enabling surface mode.
Radar mode is such a powerful and special ability i'm wondering if enabling it in survival mode should be left to optional mods?

**Comment by paramat** [comment]
Updated as suggested but i still need to test.

**Comment by paramat** [comment]
Talking of compass, i realised the minimap displays North, so i will add the crafting of a compass to this PR and add a compass to the item texture.

**Comment by TumeniNodes** [comment]
> Updated as suggested but i still need to test.
TumeniNodes i added a compass cross to your map texture, feel free to attempt this addition yourself if you want.
Talking of compass, i realised the minimap displays North, so i will add the crafting of a compass to this PR and add a compass to the item texture.

I like it paramat, looks pretty cool. Fine with that as is.

as for the compass, do you mean you will add it as an additional texture?

**Comment by paramat** [comment]
I'm thinkng of editing the texture to add a compass and pen, so that the item becomes `map:mapping_kit` .

**Comment by TumeniNodes** [comment]
You're looking to add a compass and a divider but, I think this may over-clutter the texture?
For a 16px / 16px texture this be quite difficult. Maybe either simply a separate item (either a compass or divider) might be better? You could then use it to punch a map to bring up a map-kit?

I think some brainstorming is needed here

**Comment by paramat** [comment]
Texture updated with compass and pen.
Still testing but ready for review.

**Comment by paramat** [comment]
Updated to use `minetest.after()` as requested @rubenwardy 
Retested, ready for review.

File: mods/map/init.lua (hunk 2 above)
Comment (paramat) [reply mods/map/init.lua]: Done.

File: mods/map/init.lua (hunk 1 above)
Comment (paramat) [reply mods/map/init.lua]: Done and tested without creative mod.

File: mods/map/init.lua [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+-- Global function to allow overriding
+
+function map.enable(player)
+	local creative_enabled =
+		(creative and creative.is_enabled_for(player:get_player_name())) or
+		creative_mode_cache
+
+	if creative_enabled then
+		player:hud_set_flags({minimap = true})
+		player:hud_set_flags({minimap_radar = true})
```
Comment (SmallJoker) [inline mods/map/init.lua]: These setter tables can be combined.

File: mods/map/init.lua [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+minetest.register_craftitem("map:mapping_kit", {
+	description = "Mapping Kit",
+	inventory_image = "map_mapping_kit.png",
+	stack_max = 1,
+	groups = {flammable = 3},
+	-- To instantly enable minimap
+	on_use = function(itemstack, user, pointed_thing)
+		map.enable(user)
+		return nil
```
Comment (SmallJoker) [inline mods/map/init.lua]: Superfluous return value.

File: mods/map/init.lua [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+
+-- Cache creative mode setting
+
+local creative_mode_cache = minetest.settings:get_bool("creative_mode")
+
+
+-- Global function to allow overriding
+
+function map.enable(player)
```
Comment (SmallJoker) [inline mods/map/init.lua]: Does not match with what it actually does. How about `map.update` ?

File: mods/map/init.lua (hunk 3 above)
Comment (rubenwardy) [reply mods/map/init.lua]:
```lua
local radar = creative_enabled
local minimap = creative_enabled or player:get_inventory():contains_item("main", "map:mapping_kit")
player:hud_set_flags({minimap = minimap})
player:hud_set_flags({minimap_radar = radar})
```

File: mods/map/init.lua (hunk 5 above)
Comment (rubenwardy) [reply mods/map/init.lua]: map.update_hud_flags ?

Comment (SmallJoker) [reply mods/map/init.lua]: ^ Ok too. Either function name sounds good.

Comment (paramat) [reply mods/map/init.lua]: Ok.

File: mods/map/init.lua (hunk 4 above)
Comment (paramat) [reply mods/map/init.lua]: Why? Docs say "Function must return either nil if no item shall be removed from
inventory, ..."

File: mods/map/init.lua (hunk 3 above)
Comment (paramat) [reply mods/map/init.lua]: Oh duh of course.

File: mods/map/init.lua (hunk 4 above)
Comment (paramat) [reply mods/map/init.lua]: Oh because `return nil` is the same as leaving the line out?

Comment (paramat) [reply mods/map/init.lua]: Ok well it seems to work without that line.

Comment (SmallJoker) [reply mods/map/init.lua]: Yes, paramat.

**Comment by paramat** [comment]
All done, will merge after retesting.

**Comment by jastevenson303** [comment]
I got an issue with how the client reports F9 failing with "Map is disabled by server."

Well, that's rather permanent sounding!  There's no callbacks, my issue got closed in like eighteen hours, and the new player will never know why when they set F9 to toggle map, it doesn't work in game.

Obviously, I never play in creative mode.

**Comment by paramat** [comment]
> Well, that's rather permanent sounding! There's no callbacks, my issue got closed in like eighteen hours, and the new player will never know why when they set F9 to toggle map, it doesn't work in game.

This is unclear, can you explain what your issue is?
There is a message and it is fine, so what's the issue?

**Comment by paramat** [comment]
e40f25f1fbaa206fe229002d8608f729a1d0f268

**Comment by Fixer-007** [comment]
>Well, that's rather permanent sounding! There's no callbacks, my issue got closed in like eighteen hours, and the new player will never know why when they set F9 to toggle map, it doesn't work in game.

It should work like this (from usability standpoint of long term user):
**You craft it, place it in your inventory, selecting mapkit in inventory turns on minimap immediately (there is no need for key presses or clicking anything, just select it in your hotbar as usual), changing selection to some other slot with other blocks turns off minimap immediately.**

**Comment by paramat** [comment]
Hotbar space is limited and valuable to players, better to just allow the use of minimap (F9) when the kit is somewhere in your inventory. We have F9 and shift-F9 for toggling minimap, no need to add another way to switch it on.
We plan to add inventory callbacks in the engine for instant response.

**Comment by Fixer-007** [comment]
Still feels like toyish/debuggish kind of game.

**Comment by Fixer-007** [comment]
Minimap should just showup automatically at least, that's minimum.

**Comment by paramat** [comment]
No thanks, Although that would be 'nice' it's non-essential. F9 is there for switching minimap.

**Comment by Fixer-007** [comment]
Not user friendly :)

**Comment by paramat** [comment]
No, your suggestion is user unfriendly.
Users do not want to enable minimap by selecting it on their hotbar. Because they can use F9 instead that's what they would use because it's faster and easier and doesn't take up hotbar space.
Hotbar space is limited already and players want to have other more critical items in the slots, they want to wield a weapon or tool while also being able to use minimap. Your suggestion would be a regression of behaviour and require extra callbacks and code.

**Comment by Fixer-007** [comment]
in addition to F9 is compromise

**Comment by paramat** [comment]
Very many players and server owners participate on Github in engine and MTG, the long discussions in the threads are proof of that. Here you are using this delusion to support your recent negativity about MT development. The last 2 days you have been acting like a depressed drunkard in a bar cursing everything, it's irritating, please keep it to yourself until you feel better.

**Comment by jastevenson303** [comment]
paramat is right, it needs to be a good balance in minetest_game.  Your ideas, @Fixer-007 are not without merit.  I, for example, would require the player to hold the map, and then automatically display it on HUD for them once wielded.  However, that is a sub-game, based on minetest_game.  We must be careful to distinguish between the base minetest_game, and mods and extension sub-games.

The purposes of one do not necessarily support or reflect the purposes of the other.

**Pull request closed without merging**